hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java

```diff
@@ -216,16 +217,27 @@ public void migrate(
             remainingStates.removeAll(statesToRemove);
             final var newStates = new FilteredWritableStates(writeableStates, remainingStates);
 
-            final var migrationContext =
-                    new MigrationContextImpl(previousStates, newStates, config, networkInfo, genesisRecordsBuilder);
+            // For any changes to state that depend on other services outside the current service, we need a reference
+            // to the overall state that we can pass into the context. This reference to overall state will be strictly
+            // controlled via the MigrationContext API so that only changes explicitly specified in the interface can be
+            // made (instead of allowing any arbitrary change to overall state). As above, we won't commit anything
+            // until after this service's migration
+            final var migrationStack = new SavepointStackImpl(hederaState);
```

Comment:
This looks suspicious. Changes in `writableStates` will not appear in `migrationStack` and vice versa. They are accessing different parts of the `HederaState`, but this is not enforced. Maybe we can find a solution that makes this more explicit.

Reply:
This is precisely what I was worried about, using the savepoint stack incorrectly. Let's discuss when we're both available

Reply:
Instead of using the savepoint stack, we decided to use a writable store directly to generate entity IDs. It's not pretty, but it will work for now.